Hazel.UnitTests/MessageReaderTests.cs

```diff
@@ -145,6 +145,267 @@ public void RemoveMessageWorks()
             Assert.AreEqual(Test5, five.ReadByte());
         }
 
+        [TestMethod]
+        public void InsertMessageWorks()
+        {
+            const byte Test0 = 11;
+            const byte Test3 = 33;
+            const byte Test4 = 44;
+            const byte Test5 = 55;
+            const byte TestInsert = 66;
+
+            var msg = new MessageWriter(2048);
+            msg.StartMessage(0);
+            msg.Write(Test0);
+            msg.EndMessage();
+
+            msg.StartMessage(12);
+            msg.StartMessage(23);
+
+            msg.StartMessage(34);
+            msg.Write(Test3);
+            msg.EndMessage();
+
+            msg.StartMessage(45);
+            msg.Write(Test4);
+            msg.EndMessage();
+
+            msg.EndMessage();
+            msg.EndMessage();
+
+            msg.StartMessage(56);
+            msg.Write(Test5);
+            msg.EndMessage();
+
+            MessageReader reader = MessageReader.Get(msg.Buffer);
+
+            MessageWriter writer = MessageWriter.Get(SendOption.Reliable);
+            writer.StartMessage(5);
+            writer.Write(TestInsert);
+            writer.EndMessage();
+
+            reader.ReadMessage();
+            var one = reader.ReadMessage();
+            var two = one.ReadMessage();
+            var three = two.ReadMessage();
+
+            two.InsertMessage(three, writer);
+
+            //set the position back to zero to read back the updated message
+            reader.Position = 0;
```

Comment:
I think resetting the position here is kind of unrealistic to actual usage. If you think about how we build the callstack as we read messages one, two, and three. popping all the way back to reader because we inserted in two is unlikely to happen. I would expect three to be invalid (because two has changed and two contains three) and for reading subsequent messages work normally on reader, one, and two since the head is placed after the insert.

This is important because we can't control how many inserts we may need to do or at what depth in the stack. I *think* this is automatically handled by AdjustLength, but it would be good to test it explicitly. This test suggests we can only handle one insert before having to reset everything.

Reply:
Gotcha. I had intended the reset as a way of going back and validating that we could go back and read through the whole thing after the fact, but you're right, I should be testing continuing to read as we go because that's the actual use case